Define usage of HPKE and `info` structure to be used. (#703)
openid/OpenID4VP: 1.1/openid-4-verifiable-presentations-1_1.md

Note that for the ECDH JWE algorithms (from Section 4.6 of [@!RFC7518]), the `apu` and `apv` values are inputs
into the key derivation process that is used to derive the content encryption key. Regardless of the algorithm used, the values are always part of the AEAD tag computation so will still be bound to the encrypted response.

### Encryption using HPKE

> Hybrid Public Key Encryption MAY be used by specifying the `alg` value according to [@I-D.ietf-jose-hpke-encrypt]. When it is, a `session_info` structure is calculated by the Wallet and the Verifier independently and included as mutually known private information. This is to ensure that encryption fails closed when the session information needed for verification of the credential is invalid.

Reviewer: ```suggestion
Hybrid Public Key Encryption (HPKE) may be used for encrypted responses. If HPKE is used, the `alg` value MUST be specified according to [@I-D.ietf-jose-hpke-encrypt]. Additionally, a `session_info` structure MUST be calculated and utilized for encryption and decryption independently by the Wallet and the Verifier. This ensures that decryption fails closed when the session information needed for verification of the credential is invalid. The `session_info`  structure is specific to the response modes `direct_post.jwt` and `dc_api.jwt`.
```

When the response mode is `direct_post.jwt` the encoding is as follows: 

```example

> session_info = ASCII("OpenID4VP si") ||

Reviewer 2: I know that it shouldn't really matter, but the space in `OpenID4D4VP si` makes me somewhat nervous. Would people be fine to instead use `OpenID4VP-si`?

> When the response mode is `direct_post.jwt` the encoding is as follows: 

Reviewer 3: ```suggestion
When the the presentation is invoked via redirects, such as Response Mode such as `direct_post.jwt`, the encoding is as follows: 
```

                BYTE(255) ||
                ASCII(clientId) ||
                BYTE(255) ||
                ASCII(nonce) ||
                BYTE(255) ||
                ASCII(responseUri)
```

Where:

> - ASCII("OpenID4VP-si"): A fixed ASCII string identifying this session_info structure.

Reviewer 4: ```suggestion
- ASCII("OpenID4VP-si"): A fixed ASCII string identifying this `session_info` structure.
```

> - `ASCII("OpenID4VP-si")`: A fixed ASCII string identifying this `session_info` structure.

Reviewer: Does it make sense to include "hpke" in that string?

Author: I have no strong opinion. It's a bit unnecessary though given this is only used in the context of HPKE.

- `BYTE(255)`: In each occurence is a separator byte (0xFF) used to delimit fields.

> - `ASCII(clientId)`: The `client_id` request parameter. If applicable, this includes the Client Identifier Prefix.

Reviewer: If session_info is constructed independently, it seems safer to always use Client Identifier Prefix?

```suggestion
- `ASCII(clientId)`: The `client_id` request parameter, as was passed in the Authorization Request, potentially including the Client Identifier Prefix.
```

Reviewer 3: because this is for non-DC API, but redirect based flows, there always will be client_id?

Author: But not always a prefix. This is identical to the existing session transcript text. We should be consistent with that. IMO just copying whatever value is set in client id (and so omitting the prefix of it was not included) is the least error prone solution.

'if provided' is probably a better wording than 'if applicable', but we should be consistent

- ASCII(nonce): The `nonce` request parameter.

> - ASCII(responseUri): The `redirect_uri` or `response_uri` request parameter, depending on which is present, as determined by the Response Mode.

Reviewer 4: ```suggestion
- `BYTE(255)`: In each occurence is a separator byte (0xFF) used to delimit fields.
- `ASCII(clientId)`: The `client_id` request parameter. If applicable, this includes the Client Identifier Prefix.
- `ASCII(nonce)`: The `nonce` request parameter.
- `ASCII(responseUri)`: The `redirect_uri` or `response_uri` request parameter, depending on which is present, as determined by the Response Mode.
```

> - `ASCII(responseUri)`: The `redirect_uri` or `response_uri` request parameter, depending on which is present, as determined by the Response Mode.

Reviewer: For response mode is `direct_post.jwt`, response_uri is REQUIRED, while `redirect_uri` should not be present.
```suggestion
- `ASCII(responseUri)`: The `response_uri` request parameter.
```

Author: That's a good point. While I know they aren't generally used, should we instead be specifying this for all *.jwt redirect response_modes, rather than just direct_post.jwt? (This is consistent with what we did with session transcript and why it can be one or the other there).

The following is a non-normative example of the `session_info` structure for `direct_post.jwt` response mode:

```example

> "OpenID4VP-si\xffx509_san_dns:example.com\xffexc7gBkxjx1rdc9udRrveKvSsJIq80avlXeLHhGwqtA\xffhttps://example.com/response"

Reviewer 4: ```suggestion
OpenID4VP-si\xffx509_san_dns:example.com\xffexc7gBkxjx1rdc9udRrveKvSsJIq80avlXeLHhGwqtA\xffhttps://example.com/response
```

```

The corresponding hexadecimal representation is:

```
4f70656e49443456502d7369ff783530395f73616e5f646e733a6578616d706c652e636f6dff6578633767426b786a7831726463397564527276654b7653734a4971383061766c58654c48684777717441ff68747470733a2f2f6578616d706c652e636f6d2f726573706f6e7365
```

When the response mode is `dc_api.jwt` the encoding is as follows:

```example
session_info = ASCII("OpenID4VPDCAPI-si") ||
               BYTE(255) ||
               ASCII(origin) ||
               BYTE(255) ||

>                ASCII(nonce) ||

Reviewer 4: ```suggestion
               ASCII(nonce)
```

```

Where:

> - `ASCII("OpenID4VPDCAPI-si")`: A fixed ASCII string identifying this `session_info` structure.

Reviewer: Does it make sense to include "hpke" in that string?

- `BYTE(255)`: In each occurence is a separator byte (0xFF) used to delimit fields.

> - `ASCII(origin)`: The ASCII string representing the Origin of the request as described in (#dc_api_request). It MUST NOT be prefixed with origin:.

Reviewer 5: ```suggestion
- `ASCII(origin)`: The ASCII string representing the Origin of the request as described in (#dc_api_request). It MUST NOT be prefixed with `origin:`.
```

> - ASCII(nonce): The `nonce` request parameter.

Reviewer 4: ```suggestion
- `ASCII("OpenID4VPDCAPI-si")`: A fixed ASCII string identifying this session_info structure.
- `BYTE(255)`: In each occurence is a separator byte (0xFF) used to delimit fields.
- `ASCII(origin)`: The ASCII string representing the Origin of the request as described in (#dc_api_request). It MUST NOT be prefixed with `origin:`.
- `ASCII(nonce)`: The `nonce` request parameter.
```

The following is a non-normative example of the `session_info` structure for `dc_api.jwt` response mode:

```example

> "OpenID4VPDCAPI-si\xffhttps://example.com\xffexc7gBkxjx1rdc9udRrveKvSsJIq80avlXeLHhGwqtA"

Reviewer 4: ```suggestion
OpenID4VPDCAPI-si\xffhttps://example.com\xffexc7gBkxjx1rdc9udRrveKvSsJIq80avlXeLHhGwqtA
```

```

The corresponding hexadecimal representation is:

```
4f70656e494434565044434150492d7369ff68747470733a2f2f6578616d706c652e636f6dff6578633767426b786a7831726463397564527276654b7653734a4971383061766c58654c48684777717441
```

> The `session_info` structure's bytes are used as the value of the `info` parameter when using Integrated Encryption as the Key Management Mode. If the `recipient_structure` is being used then it is used as the value of the `recipient_extra_info` parameter instead.

Reviewer: I think it makes sense to add this paragraph to the top, so readers immediately understand, where this structure is used.

   -01

>    * Add usage of HPKE an specfication for the `info` parameter. 

Reviewer 4: ```suggestion
   * Add usage of HPKE for the `info` parameter. 
```
